### PR TITLE
Implement porcelain describe

### PR DESCRIPTION
**Pull request by betesh**
I intend to implement `grep` as well, following https://stackoverflow.com/questions/15572483/how-to-do-git-grep-e-pattern-with-jgit/16263886#16263886.

It's going to be more work than `describe` was, so before I move forward, could a maintainer please tell me 1) whether the project would accept such an addition? and 2) whether it should be on a separate branch?

Thank you!

### CODE REVIEW

File: lib/rjgit.rb [hunk 1]
```diff
@@ -163,6 +163,12 @@ def self.diff(repository, options = {})
       RJGit.convert_diff_entries(diff_entries)
     end
 
+    def self.describe(repository, ref, options = {})
+      options = {:always => false, :long => false, :tags => false}.merge(options)
+      repo = RJGit.repository_type(repository)
+      git = RubyGit.new(repo).jgit
+      git.describe.set_target(ref).set_always(options[:always]).set_long(options[:long]).set_tags(options[:tags]).call
```
Comment (bartkamphorst) [inline lib/rjgit.rb:170]: This line (ending in `.call`) should be wrapped in a `begin rescue end` block and rescue the exceptions that `DescribeCommand` can throw (i.e., MissingObjectException, IncorrectObjectTypeException, RefNotFoundException, IOException).

File: lib/rjgit.rb [hunk 2]
```diff
@@ -163,6 +163,12 @@ def self.diff(repository, options = {})
       RJGit.convert_diff_entries(diff_entries)
     end
 
+    def self.describe(repository, ref, options = {})
```
Comment (bartkamphorst) [inline lib/rjgit.rb:166]: JGit's `DescribeCommand` also supports `match`. Perhaps you could enable this option as well?

File: spec/rjgit_spec.rb [hunk 3]
```diff
@@ -170,6 +170,11 @@
         end
       end
 
+    it "mimics git-describe" do
```
Comment (bartkamphorst) [inline spec/rjgit_spec.rb:173]: Please also add tests for negative cases (wrong input, or no `no names found`, etc.).

**Comment by bartkamphorst** [comment]
Regarding `grep` functionality: the gollum RJGit adapter implements grepping [here](https://github.com/gollum/rjgit_adapter/blob/master/lib/rjgit_adapter/git_layer_rjgit.rb#L179-L190) leveraging `RJGit::Porcelain.ls_tree`. @dometto Any thoughts on such an approach for RJGit itself? Or would you prefer the approach @betesh referenced in [this](https://stackoverflow.com/questions/15572483/how-to-do-git-grep-e-pattern-with-jgit) StackOverflow issue?

**Comment by dometto** [comment]
Regarding `grep`: I think I wrote that function for gollum because I thought it would be a) simpler than adding support for JGit's grep to RJGit and b) gollum also wants to match pathname, not just file contents. But support for grep in RJGit would not conflict with that, and I'm all for it! Thanks for chipping in, @betesh!

File: lib/rjgit.rb (hunk 1 above)
Comment (betesh) [reply lib/rjgit.rb:170]: how should it handle those exceptions? return nil? raise a Ruby exception of some type (what type)?

File: lib/rjgit.rb (hunk 2 above)
Comment (betesh) [reply lib/rjgit.rb:166]: Done: 7cfe77d0dda6f6a63df28543006509306f60c4a4

**Comment by betesh** [comment]
@dometto It seems like your implementation is much more straightforward?  Why not just copy-and-paste it into this repo?  (I don't mind writing a few tests, but it makes more sense to me you to push the commit and keep your name on it than for me to paste it and attribute you, especially considering that your [LICENSE](https://github.com/gollum/rjgit_adapter/blob/master/LICENSE) is empty.)

File: lib/rjgit.rb (hunk 1 above)
Comment (bartkamphorst) [reply lib/rjgit.rb]: Returning `nil` would be fine.

**Comment by dometto** [comment]
@betesh I'm fine with using the code from the adapter to implement grep, since JGit doesn't have its own `GrepCommand` anyway. @bartkamphorst how do you feel about it?

We should add a LICENSE file for sure, but I am not really worried about the attribution, especially since some lines in the snippet would have to be changed anyway (thinking of the `Gollum::Git.canonicalize` and `yield` calls for example). If you're willing to put in the time to 'port' it to RJGit and write some specs, that would be much appreciated!

**Comment by bartkamphorst** [comment]
I'm okay with using the adapter code as a basis for the actual implementation. 👍

File: lib/rjgit.rb (hunk 1 above)
Comment (betesh) [reply lib/rjgit.rb]: Done

**Comment by betesh** [comment]
@bartkamphorst Do you want to merge this and then I'll tackle grep in a separate PR?

File: lib/rjgit.rb [hunk 4]
```diff
@@ -163,6 +168,28 @@ def self.diff(repository, options = {})
...
+      options = {:always => false, :long => false, :tags => false, :match => []}.merge(options)
+      repo = RJGit.repository_type(repository)
+      git = RubyGit.new(repo).jgit
+      command = git.describe.
```
Comment (bartkamphorst) [inline lib/rjgit.rb:175]: Since we return `nil` for all exceptions, we might as well start the `begin` clause here and wrap around all the command-related calls, and have one rescue clause (rescuing IncorrectObjectTypeException, IOException, MissingObjectException, RefNotFoundException _and_ InvalidPatternException and returning `nil`).

**Comment by bartkamphorst** [comment]
> @bartkamphorst Do you want to merge this and then I'll tackle grep in a separate PR?

I have one more suggestion for a minor refactoring, but otherwise it's looking good. Tackling grep in a separate PR is the right way to go. Thanks! 👍

File: lib/rjgit.rb (hunk 4 above)
Comment (betesh) [reply lib/rjgit.rb:175]: I actually prefer not wrapping more than the part that can raise, so that the code communicates what method calls might throw what exceptions.  It's your project so I'll let you make the final call; I'm sharing my opinion just so that all the pros and cons are considered.

Comment (bartkamphorst) [reply lib/rjgit.rb:175]: Good point. This also makes sense if we want to do something more meaningful with the raised exceptions in the future.
